keps/sig-api-machinery/34-storage-hash.md

```diff
@@ -0,0 +1,181 @@
+---
+kep-number: 31
+title: Enabling clients to tell if resource endpoints serve the same set of objects
+authors:
+  - "@xuchao"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@deads2k"
+  - "@lavalamp"
+approvers:
+  - "@deads2k"
+  - "@lavalamp"
+creation-date: 2018-10-12
+last-updated: 2018-xx-xx
+status: provisional
+---
+
+# Enabling clients to tell if resource endpoints serve the same set of objects
+
+## Table of Contents
+
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+            * [Broken forward compatibility](#broken-forward-compatibility)
+            * [Inaccurate resource quota counting](#inaccurate-resource-quota-counting)
+            * [Inefficiency of API-discovery based dynamic clients](#inefficiency-of-api-discovery-based-dynamic-clients)
+      * [Goals](#goals)
+      * [Proposal](#proposal)
+         * [API changes to the discovery API](#api-changes-to-the-discovery-api)
+         * [Implementation details](#implementation-details)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+      * [Alternatives](#alternatives)
+
+## Summary
+
+We propose to expand the discovery API to enable clients to tell if resource
+endpoints (e.g., `extensions/v1beta1/replicaset` and `apps/v1/replicaset`) are
+referring the same set of objects.
+
+## Motivation
+
+Sometimes clients want to process every object in the cluster (examples below).
+Lack of visibility makes it hard for clients to understand which resources they
+need to process in that case.
+
+#### Broken forward compatibility
+
+For example, if we configure an admission webhook to validate all replicasets,
+one would set the webhook configuration this way:
+```yaml
+rules:
+  - apiGroups:
+    - "extensions,apps"
+    apiVersions:
+    - "*"
+    resources:
+    - replicasets
+```
+
+If in a future Kubernetes release, replicasets are accessible via
+`fancy-apps/v1` as well, requests sent to `fancy-apps/v1` will be left
```

Comment:
This admission webhook case is new since we last spoke about this issue and somewhat compelling.  

What would the server do with this knowledge?  Check for other registrations and send fancy-app/v1 requests through conversion and then to the extistant webhooks?  Do something else?

Reply:
Yeah, that's what I expect. The apiserver should send the converted requests to the webhooks. Defaults have already been applied to the objects before they reach the admission stage, so conversion should be possible.

Reply:
Note that since the webhook controller is part of the apiserver, so it doesn't need to go through the discovery doc to detect aliases, unless we encourage aggregated apiservers to provide aliasing endpoints for resources hosted by other apiservers.